File: src/agents/antigravity.ts
import { existsSync, rmSync } from "node:fs";
import { dirname, join } from "node:path";
import {
  addToArrayIfMissing,
  getOrCreateObject,
  readJsonFile,
  writeJsonFile,
} from "../config/json.js";
import type { Detection, RunOpts, ToolId } from "../registry.js";
import { getSkillContent } from "../tools/caveman.js";
import { verbose } from "../util/colors.js";
import { findBinaryIn } from "../util/detect.js";
import * as paths from "../util/paths.js";
import { hasOwner, removeOwner, writeOwner } from "../util/unified-block.js";

/** Detect if Antigravity is installed. */
export function detect(): Detection {
  const hasCli = !!findBinaryIn("agy", paths.antigravityKnownBinDirs());
  const hasDesktop = paths.antigravityDesktopPaths().some((p) => existsSync(p));
  if (hasCli && hasDesktop) return { installed: true, source: "cli+desktop" };
  if (hasCli) return { installed: true, source: "cli" };
  if (hasDesktop) return { installed: true, source: "desktop" };
  if (existsSync(paths.antigravityPaths().dir)) return { installed: true, source: "config" };
  return { installed: false, source: "" };
}

/** Wire a tool into Antigravity. */
export async function wire(tool: ToolId, opts: RunOpts): Promise<boolean> {
  switch (tool) {
    case "codegraph":
      wireMcp(
        "codegraph",
        paths.toksaveAbs(),
        ["runmcp", "--agent", "antigravity", "codegraph", "serve", "--mcp"],
        opts,
      );
      if (!opts.dryRun) {
        allowEntry("mcp(codegraph/*)");
        writeOwner("antigravity", "codegraph");
        installCodegraphIndexHook();
      }
      return true;
    case "context-mode":
      wireMcp(
        "context-mode",
        paths.toksaveAbs(),
        ["runmcp", "--agent", "antigravity", "context-mode"],
        opts,
      );
      if (!opts.dryRun) {
        allowEntry("mcp(context-mode/*)");
        installContextModeHook(opts);
        writeOwner("antigravity", "context-mode");
      }
      return true;
    case "rtk":
      if (!opts.dryRun) {
        allowEntry("command(rtk)");
        installRtkHook(opts);
      }
      return true;
    case "caveman":
      return wireCaveman(opts);
    case "ponytail":
      if (!opts.dryRun) writeOwner("antigravity", "ponytail");
      return true;
    case "principles":
      if (!opts.dryRun) writeOwner("antigravity", "principles");
      return true;
    default:
      return false;
  }
}

/** Unwire a tool from Antigravity. */
export async function unwire(tool: ToolId, _opts: RunOpts): Promise<boolean> {
  switch (tool) {
    case "codegraph":
      removeMcp("codegraph");
      removeOwner("antigravity", "codegraph");
      removeCodegraphIndexHook();
      return true;
    case "context-mode":
      removeMcp("context-mode");
      removeContextModeHook();
      removeOwner("antigravity", "context-mode");
      return true;
    case "rtk":
      removeRtkHook();
      return true;
    case "caveman":
      removeCaveman();
      return true;
    case "ponytail":
      removeOwner("antigravity", "ponytail");
      return true;
    case "principles":
      removeOwner("antigravity", "principles");
      return true;
    default:
      return false;
  }
}

/** Verify a tool is wired into Antigravity. */
export function verify(tool: ToolId): boolean | null {
  switch (tool) {
    case "codegraph":
      return hasMcp("codegraph");
    case "context-mode":
      return hasMcp("context-mode");
    case "rtk":
      return hasRtkHook();
    case "caveman":
      return hasCavemanSkill() || hasOwner("antigravity", "caveman");
    case "ponytail":
      return hasOwner("antigravity", "ponytail");
    case "principles":
      return hasOwner("antigravity", "principles");
    default:
      return null;
  }
}

// ─── MCP wiring (multi-surface, atomic) ─────────────────────

function wireMcp(toolId: string, command: string, args: string[], opts: RunOpts): void {
  if (opts.dryRun) return;

  verbose(`Wiring MCP ${toolId} into Antigravity (multi-surface)`, opts.verbose);

  const mcpFiles = paths.antigravityMcpFiles();
  const prepared: { file: string; cfg: Record<string, unknown> }[] = [];
  for (const f of mcpFiles) {
    paths.ensureDir(dirname(f));
    const cfg = (readJsonFile(f) as Record<string, unknown>) ?? {};
    const servers = getOrCreateObject(cfg, "mcpServers");

    const entry: Record<string, unknown> = { command };
    if (args.length > 0) entry.args = args;
    entry.trust = true;

    (servers as Record<string, unknown>)[toolId] = entry;
    prepared.push({ file: f, cfg });
  }

  atomicWriteAll(prepared, `wire MCP ${toolId}`);
}

function removeMcp(toolId: string): void {
  for (const f of paths.antigravityMcpFiles()) {
    const cfg = (readJsonFile(f) as Record<string, unknown>) ?? {};
    const mcp = cfg.mcpServers as Record<string, unknown> | undefined;
    if (mcp?.[toolId]) {
      delete mcp[toolId];
      writeJsonFile(f, cfg);
    }
  }
}

function hasMcp(toolId: string): boolean {
  for (const f of paths.antigravityMcpFiles()) {
    const cfg = (readJsonFile(f) as Record<string, unknown>) ?? {};
    const mcp = cfg.mcpServers as Record<string, unknown> | undefined;
    if (!mcp?.[toolId]) return false;
  }
  return true;
}

// ─── Permissions ─────────────────────────────────────────────

function allowEntry(entry: string): void {
  const settingsFiles = paths.antigravitySettingsFiles();
  const prepared: { file: string; cfg: Record<string, unknown> }[] = [];
  for (const f of settingsFiles) {
    paths.ensureDir(dirname(f));
    const cfg = (readJsonFile(f) as Record<string, unknown>) ?? {};
    const perms = getOrCreateObject(cfg, "permissions");
    if (!Array.isArray(perms.allow)) perms.allow = [];
    addToArrayIfMissing(perms.allow as unknown[], entry);
    prepared.push({ file: f, cfg });
  }

  atomicWriteAll(prepared, `allow ${entry}`);
}

function atomicWriteAll(
  writes: { file: string; cfg: Record<string, unknown> }[],
  operation: string,
): void {
  const originals = new Map<string, string | null>();
  for (const { file } of writes) {
    try {
      originals.set(file, paths.readFile(file));
    } catch {
      originals.set(file, null);
    }
  }

  const written: string[] = [];
  try {
    for (const { file, cfg } of writes) {
      writeJsonFile(file, cfg);
      written.push(file);
    }
  } catch (err) {
    for (const file of written) {
      const original = originals.get(file);
      if (original === null) {
        try {
          rmSync(file, { force: true });
        } catch {}
      } else if (original !== undefined) {
        try {
          paths.writeFile(file, original);
        } catch {}
      }
    }
    const msg = err instanceof Error ? err.message : String(err);
    throw new Error(`Failed to ${operation}: ${msg}`);
  }
}

// ─── RTK hook ────────────────────────────────────────────────

function installRtkHook(opts: RunOpts): void {
  const hooksFile = paths.antigravityPaths().hooks;
  paths.ensureDir(dirname(hooksFile));

  verbose("Installing RTK hook for Antigravity", opts.verbose);

  const cfg = (readJsonFile(hooksFile) as Record<string, unknown>) ?? {};
  const tok = paths.toksaveAbs();
  const command = `${tok} rtk-hook agy`;

  (cfg as Record<string, unknown>).rtk = {
    PreToolUse: [
      {
        matcher: "^(Bash|run_command|execute_command|cmd|sh|pwsh)$",
        hooks: [{ type: "command", command, timeout: 10 }],
      },
    ],
  };

  writeJsonFile(hooksFile, cfg);
}

function removeRtkHook(): void {
  const hooksFile = paths.antigravityPaths().hooks;
  const cfg = (readJsonFile(hooksFile) as Record<string, unknown>) ?? {};
  if ((cfg as Record<string, unknown>)?.rtk) {
    (cfg as Record<string, unknown>).rtk = undefined;
    writeJsonFile(hooksFile, cfg);
  }
}

function hasRtkHook(): boolean {
  const hooksFile = paths.antigravityPaths().hooks;
  const cfg = (readJsonFile(hooksFile) as Record<string, unknown>) ?? {};
  return !!(cfg as Record<string, unknown>)?.rtk;
}

// ─── Context-Mode hook ──────────────────────────────────────

function installContextModeHook(opts: RunOpts): void {
  const hooksFile = paths.antigravityPaths().hooks;
  paths.ensureDir(dirname(hooksFile));

  verbose("Installing Context-Mode hook for Antigravity", opts.verbose);

  const cfg = (readJsonFile(hooksFile) as Record<string, unknown>) ?? {};

  (cfg as Record<string, unknown>).ctx = {
    PreInvocation: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: `${paths.toksaveAbs()} context-mode-hook agy preinvocation`,
            timeout: 10,
          },
        ],
      },
    ],
    PreToolUse: [
      {
        matcher: "read_url_content|run_command|view_file",
        hooks: [
          {
            type: "command",
            command: `${paths.toksaveAbs()} context-mode-hook gemini-cli beforetool`,
            timeout: 10,
          },
        ],
      },
    ],
  };

  writeJsonFile(hooksFile, cfg);
}

function removeContextModeHook(): void {
  const hooksFile = paths.antigravityPaths().hooks;
  const cfg = (readJsonFile(hooksFile) as Record<string, unknown>) ?? {};
  if ((cfg as Record<string, unknown>)?.ctx) {
    (cfg as Record<string, unknown>).ctx = undefined;
    writeJsonFile(hooksFile, cfg);
  }
}

// ─── Codegraph index hook ───────────────────────────────────

function installCodegraphIndexHook(): void {
  const hooksFile = paths.antigravityPaths().hooks;
  paths.ensureDir(dirname(hooksFile));
  const cfg = (readJsonFile(hooksFile) as Record<string, unknown>) ?? {};
  const existing = (cfg as Record<string, unknown>).codegraphIndex as
    | Record<string, unknown>
    | undefined;
  if (existing) return;
  // Use generic hooks.json SessionStart style similar to tokless
  const hooks = getOrCreateObject(cfg, "hooks");
  const ss = (hooks.SessionStart as unknown[]) ?? [];
  const command = `${paths.toksaveAbs()} agy-hook codegraph-index`;
  if (!ss.some((g) => JSON.stringify(g).includes("codegraph-index"))) {
    ss.push({
      matcher: ".*",
      hooks: [{ type: "command", command, timeout: 10 }],
    });
    hooks.SessionStart = ss as never;
    writeJsonFile(hooksFile, cfg);
  }
}

function removeCodegraphIndexHook(): void {
  const hooksFile = paths.antigravityPaths().hooks;
  const cfg = readJsonFile(hooksFile) as Record<string, unknown> | null;
  if (!cfg) return;
  const hooks = (cfg as Record<string, unknown>)?.hooks as Record<string, unknown> | undefined;
  if (!hooks?.SessionStart) {
    // Also check top-level codegraphIndex
    if ((cfg as Record<string, unknown>).codegraphIndex) {
      delete (cfg as Record<string, unknown>).codegraphIndex;
      writeJsonFile(hooksFile, cfg);
    }
    return;
  }
  const ss = hooks.SessionStart as unknown[] | undefined;
  if (!Array.isArray(ss)) return;
  const filtered = ss.filter((g) => !JSON.stringify(g).includes("codegraph-index"));
  if (filtered.length !== ss.length) {
    if (filtered.length === 0) delete hooks.SessionStart;
    else hooks.SessionStart = filtered as never;
    writeJsonFile(hooksFile, cfg);
  }
}

// ─── Caveman ─────────────────────────────────────────────────

async function wireCaveman(opts: RunOpts): Promise<boolean> {
  if (opts.dryRun) {
    writeOwner("antigravity", "caveman");
    return true;
  }
  const gemini = paths.antigravityPaths().dir;
  const skillDir = join(gemini, "config", "skills", "caveman");
  paths.ensureDir(skillDir);
  const skillFile = join(skillDir, "SKILL.md");
  if (!existsSync(skillFile) || opts.upgrade) {
    verbose("Writing Caveman SKILL.md for Antigravity", opts.verbose);
    const skillContent = await getSkillContent();
    paths.writeFile(skillFile, skillContent);
  }
  writeOwner("antigravity", "caveman");
  return true;
}

function removeCaveman(): void {
  const gemini = paths.antigravityPaths().dir;
  const skillDir = join(gemini, "config", "skills", "caveman");
  try {
    rmSync(skillDir, { recursive: true, force: true });
  } catch {}
  removeOwner("antigravity", "caveman");
}

function hasCavemanSkill(): boolean {
  const gemini = paths.antigravityPaths().dir;
  return existsSync(join(gemini, "config", "skills", "caveman", "SKILL.md"));
}

File: rust/src/agents/antigravity.rs
use crate::agents::Agent;
use crate::registry::{Detection, RunOpts, ToolId};
use crate::util::errors::Result;
use crate::util::json::{get_or_create_object, read_json_file, write_json_file};
use crate::util::paths::{
    antigravity_known_bin_dirs, antigravity_mcp_files, antigravity_paths, toksave_abs,
};
use crate::util::unified_block::{has_owner, remove_owner, write_owner};

pub struct AntigravityAgent;

impl AntigravityAgent {
    pub fn new() -> Self {
        Self
    }
}

impl Default for AntigravityAgent {
    fn default() -> Self {
        Self::new()
    }
}

impl Agent for AntigravityAgent {
    fn detect(&self) -> Detection {
        let p = antigravity_paths();
        let has_cli = antigravity_known_bin_dirs()
            .iter()
            .any(|d| d.join("agy").exists());
        let has_config = p.dir.exists();
        if has_cli {
            Detection {
                installed: true,
                source: "cli".to_string(),
            }
        } else if has_config {
            Detection {
                installed: true,
                source: "config".to_string(),
            }
        } else {
            Detection {
                installed: false,
                source: String::new(),
            }
        }
    }

    fn wire(&self, tool: ToolId, opts: &RunOpts) -> Result<bool> {
        if opts.dry_run {
            return Ok(true);
        }
        let p = antigravity_paths();

        match tool {
            ToolId::Codegraph => {
                for mcp_file in antigravity_mcp_files() {
                    let mut cfg =
                        read_json_file(&mcp_file)?.unwrap_or_else(|| serde_json::json!({}));
                    let servers = get_or_create_object(&mut cfg, "mcpServers");
                    servers["codegraph"] = serde_json::json!({
                        "command": toksave_abs(),
                        "args": ["runmcp", "--agent", "antigravity", "codegraph", "serve", "--mcp"],
                        "trust": true
                    });
                    write_json_file(&mcp_file, &cfg)?;
                }
                write_owner("antigravity", "codegraph")?;
                Ok(true)
            }
            ToolId::ContextMode => {
                for mcp_file in antigravity_mcp_files() {
                    let mut cfg =
                        read_json_file(&mcp_file)?.unwrap_or_else(|| serde_json::json!({}));
                    let servers = get_or_create_object(&mut cfg, "mcpServers");
                    servers["context-mode"] = serde_json::json!({
                        "command": toksave_abs(),
                        "args": ["runmcp", "--agent", "antigravity", "context-mode"],
                        "trust": true
                    });
                    write_json_file(&mcp_file, &cfg)?;
                }
                write_owner("antigravity", "context-mode")?;
                Ok(true)
            }
            ToolId::Caveman => {
                write_owner("antigravity", "caveman")?;
                Ok(true)
            }
            ToolId::Rtk => {
                let mut cfg = read_json_file(&p.hooks)?.unwrap_or_else(|| serde_json::json!({}));
                cfg["rtk"] = serde_json::json!({
                    "PreToolUse": [{
                        "matcher": "^(Bash|run_command|execute_command|cmd|sh|pwsh)$",
                        "hooks": [{ "type": "command", "command": format!("{} rtk-hook agy", toksave_abs()), "timeout": 10 }]
                    }]
                });
                write_json_file(&p.hooks, &cfg)?;
                Ok(true)
            }
            ToolId::Ponytail => {
                write_owner("antigravity", "ponytail")?;
                Ok(true)
            }
            ToolId::Principles => {
                write_owner("antigravity", "principles")?;
                Ok(true)
            }
        }
    }

    fn unwire(&self, tool: ToolId, _opts: &RunOpts) -> Result<bool> {
        let p = antigravity_paths();
        match tool {
            ToolId::Codegraph => {
                for mcp_file in antigravity_mcp_files() {
                    if let Some(mut cfg) = read_json_file(&mcp_file)? {
                        if let Some(mcp) = cfg.get_mut("mcpServers").and_then(|v| v.as_object_mut())
                        {
                            mcp.remove("codegraph");
                        }
                        write_json_file(&mcp_file, &cfg)?;
                    }
                }
                remove_owner("antigravity", "codegraph")?;
                Ok(true)
            }
            ToolId::ContextMode => {
                for mcp_file in antigravity_mcp_files() {
                    if let Some(mut cfg) = read_json_file(&mcp_file)? {
                        if let Some(mcp) = cfg.get_mut("mcpServers").and_then(|v| v.as_object_mut())
                        {
                            mcp.remove("context-mode");
                        }
                        write_json_file(&mcp_file, &cfg)?;
                    }
                }
                remove_owner("antigravity", "context-mode")?;
                Ok(true)
            }
            ToolId::Caveman => {
                remove_owner("antigravity", "caveman")?;
                Ok(true)
            }
            ToolId::Rtk => {
                if let Some(mut cfg) = read_json_file(&p.hooks)? {
                    if let Some(obj) = cfg.as_object_mut() {
                        obj.remove("rtk");
                    }
                    write_json_file(&p.hooks, &cfg)?;
                }
                Ok(true)
            }
            ToolId::Ponytail => {
                remove_owner("antigravity", "ponytail")?;
                Ok(true)
            }
            ToolId::Principles => {
                remove_owner("antigravity", "principles")?;
                Ok(true)
            }
        }
    }

    fn verify(&self, tool: ToolId) -> Option<bool> {
        let p = antigravity_paths();
        match tool {
            ToolId::Codegraph => {
                let file = antigravity_mcp_files().into_iter().next()?;
                let cfg = read_json_file(&file).ok().flatten();
                Some(
                    cfg.as_ref()
                        .and_then(|c| c.get("mcpServers"))
                        .and_then(|m| m.get("codegraph"))
                        .is_some(),
                )
            }
            ToolId::ContextMode => {
                let file = antigravity_mcp_files().into_iter().next()?;
                let cfg = read_json_file(&file).ok().flatten();
                Some(
                    cfg.as_ref()
                        .and_then(|c| c.get("mcpServers"))
                        .and_then(|m| m.get("context-mode"))
                        .is_some(),
                )
            }
            ToolId::Caveman => Some(has_owner("antigravity", "caveman")),
            ToolId::Rtk => {
                let cfg = read_json_file(&p.hooks).ok().flatten();
                Some(cfg.as_ref().and_then(|c| c.get("rtk")).is_some())
            }
            ToolId::Ponytail => Some(has_owner("antigravity", "ponytail")),
            ToolId::Principles => Some(has_owner("antigravity", "principles")),
        }
    }
}
